File: ruby/hard/lru_cache.rb
class LRUCache

  =begin
    :type capacity: Integer
=end
    attr_accessor :map, :head, :tail, :capacity
    def initialize(capacity)
        @map = Hash.new
        @head = @tail = nil
        @capacity = capacity
    end


=begin
    :type key: Integer
    :rtype: Integer
=end
    def get(key)
        return -1 unless @map.has_key?(key)
        node = @map[key]
        move_node_to_front(node)
        return node.value
    end


=begin
    :type key: Integer
    :type value: Integer
    :rtype: Void
=end
    def put(key, value)
        if @map.has_key?(key)
            node = @map[key]
            node.value = value
            move_node_to_front(node)
        else
            node = ListNode.new(key, value)
            insert_node_at_front(node)
        end
    end
    
    def move_node_to_front(node)
        return if(node == @head || @head == @tail)
        node.prev.next = node.next
        if(node == @tail)
            @tail = @tail.prev
            @tail.next = nil
        else
            node.next.prev = node.prev
        end
        node.prev = nil
        push(node)
    end
    
    def insert_node_at_front(node)
        if @map.size >= capacity
            @map.delete(@tail.key)
            pop
        end
        @map[node.key] = node
        push(node)
    end

    def push(node)
        if @head.nil? && @tail.nil?
            @head = @tail = node
        else
            node.next = @head
            @head.prev = node
            @head = node
        end
    end
    
    def pop
        if @head == @tail
            @head = @tail = nil
        else
            @tail = @tail.prev
            @tail.next = nil
        end
    end
end

class ListNode
    attr_accessor :key, :value, :next, :prev
    def initialize(key, value)
        @key = key
        @value = value
    end
end

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache.new(capacity)
# param_1 = obj.get(key)
# obj.put(key, value)

File: javascript/hard/lRUCache.js
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.map = new Map();
    this.capacity = capacity;
    this.head = null;
    this.tail = null;
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if(!this.map.has(key)) return -1;
    let node = this.map.get(key);
    this.moveNodeToFront(node);
    return node.value;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    if(this.map.has(key)) {
        let node = this.map.get(key);
        node.value = value;
        this.moveNodeToFront(node);
    } else {
        let node = new ListNode(key, value);
        this.insertNodeToFront(node);
    }
};

LRUCache.prototype.moveNodeToFront = function(node) {
    if(this.head === this.tail || node === this.head) return;
    node.prev.next = node.next;
    if(node === this.tail) {
        this.tail = node.prev;
    } else {
        node.next.prev = node.prev;
    }
    this.push(node);
};

LRUCache.prototype.insertNodeToFront = function(node) {
    if(this.map.size >= this.capacity) {
        this.map.delete(this.tail.key);
        this.pop();
    }
    this.map.set(node.key, node);
    this.push(node);
};

LRUCache.prototype.push = function(node) {
    if(this.head === null && this.tail === null) {
        this.head = this.tail = node;
    } else {
        node.next = this.head;
        this.head.prev = node;
        this.head = node;
    }
};

LRUCache.prototype.pop = function() {
    if(this.head === this.tail) {
        this.head = this.tail = null;
    } else {
        this.tail = this.tail.prev;
        this.tail.next = null;
    }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = Object.create(LRUCache).createNew(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */

var ListNode = function(key, value) {
    this.key = key;
    this.value = value;
    this.next = null;
    this.prev = null;
};
